gazebo_plugins/src/gazebo_ros_ray_sensor.cpp

```diff
@@ -0,0 +1,428 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gazebo_plugins/gazebo_ros_ray_sensor.hpp>
+#include <gazebo/transport/transport.hh>
+#include <sensor_msgs/point_cloud2_iterator.hpp>
+#include <gazebo_ros/conversions.hpp>
+
+#include <string>
+#include <algorithm>
+#include <limits>
+#include <memory>
+
+namespace gazebo_plugins
+{
+
+class GazeboRosRaySensorPrivate
+{
+public:
+  /// Node for ROS communication.
+  gazebo_ros::Node::SharedPtr rosnode_;
+
+  /// Sensor this plugin is attached to
+  boost::variant<gazebo::sensors::RaySensorPtr, gazebo::sensors::GpuRaySensorPtr> sensor_;
+
+  // Aliases
+  using LaserScan = sensor_msgs::msg::LaserScan;
+  using PointCloud = sensor_msgs::msg::PointCloud;
+  using PointCloud2 = sensor_msgs::msg::PointCloud2;
+  using Range = sensor_msgs::msg::Range;
+  using LaserScanPub = rclcpp::Publisher<LaserScan>::SharedPtr;
+  using PointCloudPub = rclcpp::Publisher<PointCloud>::SharedPtr;
+  using PointCloud2Pub = rclcpp::Publisher<PointCloud2>::SharedPtr;
+  using RangePub = rclcpp::Publisher<sensor_msgs::msg::Range>::SharedPtr;
+
+  /// Publisher of output
+  /// \todo use std::variant one c++17 is supported in ROS2
+  boost::variant<LaserScanPub, PointCloudPub, PointCloud2Pub, RangePub> pub_;
```

Comment:
I decided to use variant because I saw it was already needed for  [sdformat](https://bitbucket.org/osrf/sdformat/src/13546ead54939168420eb770497d189459da6195/include/sdf/Param.hh#lines-219) so I believe it doesn't represent an additional dependency.

Reply:
Ok sounds good. We've been trying to stop using boost, but you're right that it is already a dependency.